promql/engine.go

```diff
@@ -884,6 +880,14 @@ func (ev *evaluator) eval(expr Expr) Value {
 		}
 
 		sel := e.Args[matrixArgIndex].(*MatrixSelector)
+		if sel.series == nil {
```

Comment:
So I had to change this around from what I originally did because it didn't quite work. Originally I just had the unexpanded series set stored in the selector and expanded it when I needed to use it. But this became an issue when doing a range query because it tries to expand the series set multiple times, which is a problem because once the series is expanded once it "uses up" the iterator. I don't know how to phrase that better, but the second time it tries to expand it the iterator has already gone through to the end. So instead, after I expand it the first time, I store it in sel.series and reuse that value. So it should be nil the first time through but only the first time through.

Reply:
Okay. This is repeated a few times, it should probably be factored out